### PR TITLE
Use an IndexVec to cache queries with index-like key

### PR DESCRIPTION
**Pull request by cjgillot**
Revival of an old idea. Let's see if it has more effect.

r? @ghost

### CODE REVIEW

File: compiler/rustc_middle/src/query/keys.rs [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
-use rustc_middle::ty::subst::{GenericArg, SubstsRef};
-use rustc_middle::ty::{self, layout::TyAndLayout, Ty, TyCtxt};
+use rustc_query_system::query::{DefaultCacheSelector, VecCacheSelector};
 use rustc_span::symbol::{Ident, Symbol};
 use rustc_span::{Span, DUMMY_SP};
 
 /// The `Key` trait controls what types can legally be used as the key
 /// for a query.
 pub trait Key {
+    type CacheSelector;
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/query/keys.rs:18]:
```suggestion
    type CacheSelector = DefaultCacheSelector<Self>;
```
Seems reasonable to set a default here.

**Comment by cjgillot** [comment]
r? compiler

**Pull request merged**